archive/imo/imo2008_q2.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import data.real.basic
+import data.set.basic
+import data.set.finite
+
+/-!
+# IMO 2008 Q2
+(a) Prove that
+          ```x^2 / (x-1)^2 + y^2 / (y-1)^2 + z^2 / (z-1)^2 ≥ 1```
+for all real numbers x, y, z, each different from 1, and satisfying xyz = 1.
+
+(b) Prove that equality holds above for infinitely many triples of rational numbers x, y, z, each
+different from 1, and satisfying xyz = 1.
+
+# Solution
+(a) Since xyz = 1, we can apply the substitution x = a/b, y = b/c, z = c/a.
+Then we define m = c-b, n = b-a and rewrite the inequality as LHS - 1 ≥ 0
+using c, m and n. We factor LHS - 1 as a square, which finishes the proof.
+
+(b) We present a set W of rational triples. We prove that W is a subset of the
+set of rational solutions to the equation, and that W is infinite.
+-/
+
+lemma subst_abc {x y z : ℝ} (h : x*y*z = 1) :
+  ∃ a b c : ℝ, a ≠ 0 ∧ b ≠ 0 ∧ c ≠ 0 ∧ x = a/b ∧ y = b/c ∧ z = c /a :=
+begin
+  use [x, 1, 1/y],
+  have h₁ : x ≠ 0,              exact left_ne_zero_of_mul (left_ne_zero_of_mul_eq_one h),
```

Comment:
All of these `have ..., exact`s here and below can be removed by just giving the term after `:=`.
```suggestion
  have h₁ : x ≠ 0 := left_ne_zero_of_mul (left_ne_zero_of_mul_eq_one h),
```

Reply:
done!